docs/DesignDocs/SGX_Endorsements_V2.md

```diff
@@ -0,0 +1,222 @@
+
+### Get TCB Info
+1. tcbInfo:version changed from 1 to 2.
+2. New field tcbInfo:tcbType.
+3. New field tcbInfo:tcbEvaluationDataNumber.
+4. Field tcbInfo:tcbLevels:status was renamed to tcbInfo:tcbLevels:tcbStatus.
+5. New fields in tcbInfo:tcbLevels named tcbDate and advisoryIDs.
+
+### Get QE Identity Info
+1. The 'qeIdentity' block was renamed to 'enclaveIdentity'.
+2. A new field, enclaveIdentity:id was added to differentiate QE Id info from Quoting Enclave (QE) and Quoting Validation Enclave (QVE).
+3. enclaveIdentity:version changed from 1 to 2.
+4. New field enclaveIdentity:tcbEvaluationDataNumber.
+5. qeIdentity:isvsvn moved to enclaveIdentity:tcbLevels:tcb:isvsvn.
+6. New fields in enclaveIdentity:tcbLevels named tcbData, tcbStatus and advisoryIDs.
+
+### New field definitions
+
+#### tcbType:
+	type: integer
+	example: 0
+	description: >- Type of TCB level composition that determines TCB level comparison logic
+
+##### tcbEvaluationDataNumber:
+	type: integer
+	example: 2
+	description: >- A monotonically increasing sequence number changed when Intel updates the content of the TCB evaluation data set: TCB Info, QE Idenity and QVE Identity. The tcbEvaluationDataNumber update is synchronized across TCB Info for all flavors of SGX CPUs (Family-Model-Stepping-Platform-CustomSKU) and QE/QVE Identity. This sequence number allows users to easily determine when a particular TCB Info/QE Idenity/QVE Identiy superseedes another TCB Info/QE Identity/QVE Identity (value: current TCB Recovery event number stored in the database).
+
+##### tcbDate:
+	type: string
+	format: date-time
+	description: >- Representation of date and time when the TCB level was certified not to be vulnerable to any issues described in SAs that were published on or prior to this date. The time shall be in UTC and the encoding shall be compliant to ISO 8601 standard (YYYY-MM-DDThh:mm:ssZ).
+
+##### tcbStatus:
+	type: string
+	description: TCB level status
+
+##### advisoryIDs:
+	type: array
+	description: >- Array of Advisory IDs describing vulnerabilities that this TCB level is vulnerable to. Note: The value can be different for different FMSPCs. This field is optional. It will be present only if the list of Advisory IDs is not empty.
+	items:
+        type: string
+
+For more information on the new fields please see [API Version 2](https://api.portal.trustedservices.intel.com/documentation#pcs-certificate-v2).
+
+
+## Changes
+
+### Update current structs with new additional fields and create new type definitions.
+
+```C
+typedef enum _oe_tcb_level_status
+{
+    // Existing fields
+    OE_TCB_LEVEL_STATUS_UNKNOWN,
+    OE_TCB_LEVEL_STATUS_REVOKED,
+    OE_TCB_LEVEL_STATUS_OUT_OF_DATE,
+    OE_TCB_LEVEL_STATUS_CONFIGURATION_NEEDED,
+    OE_TCB_LEVEL_STATUS_UP_TO_DATE,
+
+    // New field for new value "OutOfDateConfigurationNeeded"
+    OE_TCB_LEVEL_STATUS_OUTOFDATE_CONFIGURATION_NEEDED,
+
```

Comment:
This value is kind of confusing. 
I am wondering: would the status be considered as up-to-date if configured properly?

Reply:
I think the status "OutOfDateConfigurationNeeded" is a subset of "OutOfDate", so it will not be considered "UpToDate".

Reply:
Agree that this value is confusing for the above reasons.  In my opinion, the names should be clear, and statuses should be disjoint, so that one status is not a subset of another status, if this is an enum.  My guess though is that you mean that OE_TCB_LEVEL_STATUS_CONFIGURATION_NEEDED and OE_TCB_LEVEL_STATUS_OUT_OF_DATE are both set, and so you need a separate value.   If so, then it seems to me that the use of an enum is itself problematic and replacing it with a bitfield would be better.

Reply:
I found some logic in the Intel github that sheds some more information on this new status and looks like a combination of both the QE Identity Info status and the TCB Info status.  For now I will keep it as a separate status given that this is its own special case.

Status convergeTcbstatus(Status tcbInfoStatus, Status qeIdentityStatus) {
    if (qeIdentityStatus == STATUS_OK) {
        return tcbInfoStatus;
    }
    if (qeIdentityStatus == STATUS_SGX_ENCLAVE_IDENTITY_OUT_OF_DATE ||
        qeIdentityStatus == STATUS_SGX_ENCLAVE_REPORT_ISVSVN_OUT_OF_DATE ||
        qeIdentityStatus == STATUS_QE_IDENTITY_OUT_OF_DATE) {
        switch (tcbInfoStatus)
        {
        case STATUS_OK:
            return STATUS_TCB_OUT_OF_DATE;
        case STATUS_TCB_CONFIGURATION_NEEDED:
            return STATUS_TCB_OUT_OF_DATE_CONFIGURATION_NEEDED;
        default:
            return tcbInfoStatus;
        }
    }
    return STATUS_INVALID_PARAMETER;
}

Reply:
To me that code confirms that an enum is not appropriate since it's really an orthogonal property.  Do we really need to match a bad API design here?

Reply:
I agree that using the bitfield is better, but I am not 100% sure what is the original intent of this new status.

Reply:
Ok, let's change it to a bitfield.  It is definitely less confusing.